daml-foundations/daml-ghc/ide/src/Development/IDE/State/Rules/Daml.hs

```diff
@@ -339,6 +335,23 @@ ofInterestRule = do
               prevCtxRoots <- modifyVar envPreviousScenarioContexts $ \prevCtxs -> pure (ctxRoots, prevCtxs)
               when (prevCtxRoots /= ctxRoots) $ void $ SS.gcCtxs scenarioService ctxRoots
 
+getFilesOfInterestRule :: Rules ()
+getFilesOfInterestRule = do
+    defineEarlyCutoff $ \GetFilesOfInterest _file -> assert (null _file) $ do
+        alwaysRerun
+        Env{..} <- getServiceEnv
+        filesOfInterest <- liftIO $ readVar envOfInterestVar
+        pure (Just $ T.encodeUtf8 $ T.pack $ show filesOfInterest, ([], Just filesOfInterest))
+
+getOpenVirtualResourcesRule :: Rules ()
+getOpenVirtualResourcesRule = do
+    defineEarlyCutoff $ \GetOpenVirtualResources _file -> assert (null _file) $ do
+        alwaysRerun
+        DamlEnv{..} <- getDamlServiceEnv
+        openVRs <- liftIO $ readVar envOpenVirtualResources
+        pure (Just $ T.encodeUtf8 $ T.pack $ show openVRs, ([], Just openVRs))
```

Comment:
Are you looking for Data.ByteString.UTF8.fromString?

Reply:
I was trying to avoid adding the dependency on `utf8-string` but 1. we already depend on it transitively and 2. there is really no reason for doing that so changed to `Data.ByteString.UTF8.fromString`.

Reply:
At one point I spent some time optimising it, specifically this function, so keen to make use of it 😄